urld0f681c558_farom57_AstroidDriver_tgz-pJ8-laazotea_indi_server_examples_MinimumINDIServerJ8/src/laazotea/indi/server/examples/INDIBasicServer.java

```diff
@@ -0,0 +1,526 @@
+/*
+ *  This file is part of INDI for Java Server.
+ * 
+ *  INDI for Java Server is free software: you can redistribute it
+ *  and/or modify it under the terms of the GNU General Public License 
+ *  as published by the Free Software Foundation, either version 3 of 
+ *  the License, or (at your option) any later version.
+ * 
+ *  INDI for Java Server is distributed in the hope that it will be
+ *  useful, but WITHOUT ANY WARRANTY; without even the implied warranty
+ *  of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ * 
+ *  You should have received a copy of the GNU General Public License
+ *  along with INDI for Java Server.  If not, see 
+ *  <http://www.gnu.org/licenses/>.
+ */
+package laazotea.indi.server.examples;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import laazotea.indi.INDIException;
+import laazotea.indi.server.DefaultINDIServer;
+import laazotea.indi.server.INDIClient;
+import laazotea.indi.server.INDIDevice;
+
+/**
+ * A simple INDI Server that basically sends all messages from drivers and
+ * clients and viceversa, just performing basic checks of messages integrity.
+ *
+ * It allows to dinamically load / unload different kinds of Devices with simple
+ * shell commands.
+ *
+ * @author S. Alonso (Zerjillo) [zerjioi at ugr.es]
+ * @version 1.32, July 23, 2013
+ */
+public class INDIBasicServer extends DefaultINDIServer {
+
+  /**
+   * The only server.
+   */
+  private static INDIBasicServer server;
+  /**
+   * Loaded jar files.
+   */
+  private static ArrayList<String> jarFiles;
+
+  /**
+   * Constructs the server.
+   */
+  public INDIBasicServer() {
+    super();
+
+    jarFiles = new ArrayList<String>();
+  }
+
+  /**
+   * Constructs the server with a particular port.
+   *
+   * @param port The port to which the server will listen.
+   */
+  public INDIBasicServer(int port) {
+    super(port);
+
+    jarFiles = new ArrayList<String>();
+  }
+
+  /**
+   * Loads the Java Drivers in a JAR file
+   *
+   * @param jar The JAR file
+   * @see #unloadJava
+   */
+  public void loadJava(String jar) {
+    try {
+      loadJavaDriversFromJAR(jar);
+
+      jarFiles.add(jar);
+    } catch (INDIException e) {
+      System.err.println(e.getMessage());
+    }
+  }
+
+  /**
+   * Unloads the Java Drivers in a JAR file.
+   *
+   * @param jar The JAR file.
+   * @see #loadJava
+   */
+  public void unloadJava(String jar) {
+    destroyJavaDriversFromJAR(jar);
+
+    jarFiles.remove(jar);
+  }
+
+  private void reloadJava(String jar) {
+    server.unloadJava(jar);
+
+    try {
+      Thread.sleep(500);
+    } catch (InterruptedException e) {
+    }
+
+    while (server.isAlreadyLoaded(jar)) {
+      System.err.println("Waiting for " + jar + " to unload");
+      try {
+        Thread.sleep(500);
+      } catch (InterruptedException e) {
+      }
+    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 106:112, 446:452.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.